include/mpscringbuffer.hpp

```diff
@@ -0,0 +1,52 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+#pragma once
+
+#include <atomic>
+#include <chrono>
+#include <thread>
+
+namespace ddprof {
+//  mimic: std::hardware_destructive_interference_size, C++17
+constexpr std::size_t hardware_destructive_interference_size = 128;
+
+struct MPSCRingBufferMetaDataPage {
+  alignas(hardware_destructive_interference_size) uint64_t writer_pos;
+  alignas(hardware_destructive_interference_size) uint64_t reader_pos;
+  alignas(hardware_destructive_interference_size) std::atomic_flag spinlock;
+};
+
+inline void lock(std::atomic_flag *lock) {
+  static constexpr uint32_t k_max_active_spin = 4000;
+  static constexpr std::chrono::nanoseconds k_yield_sleep =
+      std::chrono::microseconds(500);
+
+  uint32_t spincount = 0;
+
+  for (;;) {
+    if (!lock->test_and_set(std::memory_order_acquire)) {
+      break;
+    }
+    do {
+      if (spincount < k_max_active_spin) {
+        ++spincount;
+#ifdef __x86_64__
+        asm volatile("pause");
+#else
+        asm volatile("yield");
+#endif
+      } else {
+        // If active spin fails, yield
```

Comment:
Since this spinlock is also supposed to support multi-process coordination, I think we might need to make it more robust.  Here are some random ideas.

1.  As Ivo notes in the other comment, when `n_waiters >> n_schedulable_cores`, then it becomes more likely for the thread holding the lock to slow down in a way that compounds contention.  The way that some of the pthreads types handle this is by having the slow-path do an atomic increment indicating the number of "interested" (haha) threads.  Unfortunately, the pthreads code I checked only uses this field for arithmetic, but I guess we could use a combination of the lock-defined `num_interested` and thread-local `interested_idx` to sequence the waits?  This isn't a complete solution, just an idea.
2.  As Erwan indicated, since we're at the mercy of other processes, which may be killed in various situations (user culling heavy processes, OOM, or even in the case where the lock is held across container boundaries and an unrelated service gracefully closes), I wonder if we need to structure a check during the slow path.  This is actually really hard to do because the lock-holder may be in a separate PID namespace.  I'd say that after a certain number of passes (plus the addition of a sequence number or something, so we don't get a lot of threads thrashing on the lock), one of the threads needs to make a call to either take ownership of the lock or declare the lock invalid so the application can figure out what to do.

Lots of words with nothing specific, sorry :(

Reply:
Concerning point 1, I will add a benchmark to be able to base our decisions on measures.
Point 2 is very important, I definitely need to add safeguards. Disabling allocation profiling when we detect that a lock is never released seems the safest choice to me.